### DISCUSSION TITLE
Speculations from the wee small hours

### DISCUSSION THREAD

**Discussion by FullValueRider** [twinbasic/twinbasic#667] [General]
I'm not a professional programmer but I do try to think about what I'd trying to achieve and the best way to do so..

It came to me in the wee small hours over the weekend that perhaps programming languages are doing Types wrong.

It seemed to me that there is only one type available for current microprocessors and that that type is the bit (representing 0 or 1).  All current types are thus collections of bits which are currently organised on the basis of logical structure of the processor.

If a programming language were to implement the Bit as the one Type built into the compiler it seems to me that the logical next step is that the collections of bits that are used by current programming languages become a Library issue rather than a compiler issue.

e.g. (with dubious syntax)
Type Integer As Bit(0 to 31)
Type Byte as Bit(0 to 7)
Type Char as Bit(0 to 15)
Type Single as Bit(Sign(63),Exponent(53 to 62),Significant(0 to 52)
etc.

I thought that was an interesting idea and wished I had the knowledge to be able to look at that idea from a compiler design perspective and whether it would have any advantages.

I then got to speculating (based on the discussion of Bit manipulation that we've had) that types based on collections of Bits that extending the To 'operator' in twinBasic to be a more general slice operator would allow

Dim myVar1 as long = <some value>
Dim myVar2 as long = <some value>

myVar1(3 to 6) = myVar2(1 to 3) (i.e. Bits 1 to 3 of myVar2 are mapped to Bits (3 to 6) of myVar1.  

But as I say, just idle speculation.

Feel free to follow on with your own daydreams and speculations.

Cheers

**Comment by bclothier** [comment]
There are few more considerations, however.

1) It's usually faster to work with words, not bits. If your processor is 64-bit, you are going to be loading 64-bits of data into registers whether you want it or not. Therefore, the "word" which is 64-bit in this case<sup>1</sup>, so it's actually better to work with 64-bit integers on 64-bit processors but 32-bit integers on 32-bit processors.

2) As the consequence of the 1) there is also memory alignment to consider. Consider:

```
Public Type A
  v As Long
  w As Integer
  y As Double
  x As Integer
  z As Double
End Type

Public Type B
  w As Integer
  y As Double
  v As Long
  z As Double
  x As Integer
End Type

Public Type C
  w As Integer
  x As Integer
  v As Long
  y As Double
  z As Double
End Type

Public MyA As A, MyB As B, MyC As C
```

And via immediate windows (ran on a 64-bit VBA):
```
Debug.Print LenB(MyA), LenB(MyB), LenB(MyC)
 32            40            24 
```

Note that simply by rearranging we got different sized structures. Turns out that the order of the members affect whether it's going to fit into the word boundary -- if it can't, the compiler will push it out to the next word boundary, which mean the remaining bits within that word boundary is left unused. This is done usually because it's easier to optimize code when they stay aligned to a word boundary so wasting a few more bits is acceptable, unless you write explicit code to pack the structure and doing so usually is slower.

3) Not all encoding are equal. In VBx, we are using ucs-2 encoding, which is why you only ever see a `String`, and none of that muck that C/C++ people have to go through with C-strings, wide-strings, DBCS, or several other variants. I'll let [Joel Spolsky discuss this problem a bit more](https://www.joelonsoftware.com/2003/10/08/the-absolute-minimum-every-software-developer-absolutely-positively-must-know-about-unicode-and-character-sets-no-excuses/). Though VBx lets us play pretend that strings are just byte arrays, we better respect the encoding:

```
Dim s As String
s = "Hello, world!"
Dim b() As Byte
b = s
b(1) = &H10
b(3) = &H20
b(5) = &H40
s = b
Debug.Print s 'the first 3 letters are now malformed, printing "?"
```

Likewise:
```
Debug.Print StrConv(s, vbFromUnicode) 'Prints gibberish; the string was re-encoded into ASCII
```

To demonstrate how it's encoding the contents, you can do:
```
Dim s As String
s = "Hello, world!"

Dim b1() As Byte
b1 = s
Dim b2() As Byte
b2 = StrConv(s, vbFromUnicode)

Dim i As Long
For i = LBound(b1) To UBound(b1)
  Debug.Print b1(i), ;
Next
Debug.Print vbNullString
For i = LBound(b2) To UBound(b2)
  Debug.Print b2(i), ;
Next
```

But as Joel explains, Unicode gets really complicated. By using surrogate characters, you can generate gli̵t̶c̸hy tę̴̒x̸̞̓t̶͖͝. That requires more bits to encode those properly.  For that reason, we usually don't work with bits.

That said, VBx already has a way to do a direct byte copying without using any API but it takes a bit of work.

```
Public Type MyCurrency
  Value As Currency
End Type

Public Type MyDouble
  Value As Double
End Type

Public MyC As MyCurrency, MyD As MyDouble

...

MyD.Value = 1.0
LSet MyC = MyD
Debug.Print MyC.Value
MyC.Value = MyC.Value + 1
LSet MyD = MyC
Debug.Print MyD.Value
```

The result of the `MyD.Value` will be `1.00000000000222 `, which happens to be the very next value that can be encoded into `Double` without rounding up/down above the value `1.0`. With that setup, you can then set up a ULPS test which can be superior compared to the more usual epsilon test (e.g. `lft - rgt > 0.001`).

So even though we don't directly work with bits, we still can do by manipulating the types (usually the integral types as they are much easier to work with). Because of extra complexity around the bit twiddling, we usually don't bother with that level of details lest we risk violating the abstraction that the data types give us and generate undefined behaviors. 

----

<sup>1. Technically, while the size of a word should align to the processor's size, there are still literatures where it may not align to the processor's actual size. For example, Win32 APIs all use the name `DWORD` to refer to a 32-bit integers, and `WORD` to refer to a 16-bit integers, an artifact of the old 16-bit computing which they could not change without breaking the APIs. For that reason, `WORD` and `DWORD`s are in fact _fixed-size_ even though the name "word" originally wasn't meant to refer to a fixed size. So be careful about how that word is used.

**Comment by FullValueRider** [comment]
Wow, I never expected responses to my idle thinking.  However, your response does touch upon the 'problem' as I see it.  You've made the underlying architecture of the processor a property of the compiler so your thinking would appear to be already constrained because now you have to couch new types in the context of the underlying process architecture.

I think,  in my idle speculation, the compiler can know about/discover the architecture of the target, i.e. how collections of bits are organised on its target processor, and thus organise the code it generates in line with the processor architecture.  So the generated code would still be working in terms of machine words, no los of efficiency there, but it does mean that you are now free to design  'types' at a more fundamental level.

**Comment by VBFNiya** [comment]
This is just my personal opinion but I think adding a Bit data type is a bit(no pun intended) overkill. We can work with bits just fine using normal everyday bitwise operators. However, I don't feel strongly enough to say I'd oppose the addition of a Bit type either. It certainly can't hurt.

**Comment by FullValueRider** [comment]
It's not the intent of this discussion to request a Bit Type for twinBasic.  I was speculating more on breaking the link between Types and microprocessor logic.  i.e. in my mythical language I could define a signed integer which was 197 bits long (if I wanted to).  It would then be up to the compiler to work out how to map that definition to the underlying machine structure.  

Its a similar type of thinking to that which gave rise to the Uniform Function Call Syntax (see https://en.wikipedia.org/wiki/Uniform_Function_Call_Syntax) in the D and Nim programming languages.
